src/auth_context.rs

```diff
@@ -0,0 +1,155 @@
+};
+
+/// To perform server-side authentication, gRPC exposes the authentication context
+/// for each call. The context exposes important authentication-related information
+/// about the RPC such as the type of security/authentication type being used and
+/// the peer identity.
+///
+/// The authentication context is structured as a multi-map of key-value pairs -
+/// the auth properties. In addition to that, for authenticated RPCs, the set of
+/// properties corresponding to a selected key will represent the verified identity
+/// of the caller - the peer identity.
+///
+/// The contents of the auth properties are populated by an auth interceptor within
+/// gRPC Core. The interceptor also chooses which property key will act as the peer
+/// identity (e.g. for client certificate authentication this property will be
+/// `x509_common_name` or `x509_subject_alternative_name`).
+pub struct AuthContext {
+    ctx: *mut grpc_auth_context,
+}
+
+/// Binding to gRPC Core AuthContext
+impl AuthContext {
+    pub(crate) unsafe fn from_call_ptr(call: *mut grpc_call) -> Option<Self> {
+        let ctx = grpc_sys::grpc_call_auth_context(call);
+        if ctx.is_null() {
+            None
+        } else {
+            Some(AuthContext { ctx })
+        }
+    }
+
+    /// The name of the property gRPC Core has chosen as main peer identity property,
+    /// if any.
+    pub fn peer_identity_property_name(&self) -> Option<&str> {
+        assert!(!self.ctx.is_null());
+        unsafe {
+            let p = grpc_sys::grpc_auth_context_peer_identity_property_name(self.ctx);
+            if p.is_null() {
+                None
+            } else {
+                Some(CStr::from_ptr(p).to_str().expect("valid UTF-8 data"))
+            }
+        }
+    }
+
+    /// `true` if the client has provided a valid certificate (or other auth method
+    /// considered valid by gRPC).
+    /// `false` in non-secure scenarios.
+    pub fn peer_is_authenticated(&self) -> bool {
+        assert!(!self.ctx.is_null());
+        unsafe { grpc_sys::grpc_auth_context_peer_is_authenticated(self.ctx) != 0 }
+    }
+
+    /// `AuthContext[peer_identity_property_name()]`
+    ///
+    /// There may be several of them (for instance if `x509_subject_alternative_name` is selected)
+    pub fn peer_identity(&self) -> AuthPropertyIter {
+        assert!(!self.ctx.is_null());
+        unsafe {
+            // grpc_auth_context_peer_identity returns empty_iterator when self.ctx is NULL
+            let iter = grpc_sys::grpc_auth_context_peer_identity(self.ctx);
+            AuthPropertyIter {
+                iter,
+                _lifetime: PhantomData,
+            }
+        }
+    }
+}
+
+impl<'a> IntoIterator for &'a AuthContext {
+    type Item = AuthProperty<'a>;
+    type IntoIter = AuthPropertyIter<'a>;
+
+    /// Iterate over the AuthContext properties
+    fn into_iter(self) -> Self::IntoIter {
+        unsafe {
```

Comment:
Why not just `self.peer_identity()` or `AuthContext::peer_identity(self)`?

Reply:
I don't see what you mean here. :/

Reply:
`grpc_auth_context_property_iterator` iterates over more elements than `grpc_auth_context_peer_identity` does.

Reply:
`IntoIterator::into_iter` is the same as `AuthContext::peer_identity`.

Reply:
Hmm did you see my last message ? I think we posted it at the same time.
https://github.com/grpc/grpc/blob/695c844d0fc5ac216a281418fa521130807fa31f/src/core/lib/security/context/security_context.cc#L214-L220

Reply:
Got it, thanks!